// vxe-pc-ui@4.1.2/lib/loading/src/loading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _ui = require("../../ui");
var _vn = require("../../ui/src/vn");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = exports.default = (0, _vue.defineComponent)({
  name: 'VxeLoading',
  props: {
    modelValue: Boolean,
    icon: {
      type: String,
      default: () => (0, _ui.getConfig)().loading.icon
    },
    text: {
      type: String,
      default: () => (0, _ui.getConfig)().loading.text
    },
    status: String
  },
  setup(props, {
    slots
  }) {
    const refInitialized = (0, _vue.ref)(false);
    const computeLoadingIcon = (0, _vue.computed)(() => {
      return props.icon || (0, _ui.getIcon)().LOADING;
    });
    const computeLoadingText = (0, _vue.computed)(() => {
      const {
        text
      } = props;
      return _xeUtils.default.isString(text) ? text : (0, _ui.getI18n)('vxe.loading.text');
    });
    const handleInit = () => {
      if (!refInitialized.value) {
        refInitialized.value = !!props.modelValue;
      }
    };
    (0, _vue.watch)(() => props.modelValue, () => {
      handleInit();
    });
    handleInit();
    return () => {
      const {
        modelValue,
        status
      } = props;
      const defaultSlot = slots.default;
      const textSlot = slots.text;
      const iconSlot = slots.icon;
      const initialized = refInitialized.value;
      const loadingIcon = computeLoadingIcon.value;
      const loadingText = computeLoadingText.value;
      if (!initialized && !modelValue) {
        return (0, _vue.createCommentVNode)();
      }
      return (0, _vue.h)('div', {
        class: ['vxe-loading', {
          [`theme--${status}`]: status,
          'is--visible': modelValue
        }]
      }, defaultSlot ? [(0, _vue.h)('div', {
        class: 'vxe-loading--wrapper'
      }, (0, _vn.getSlotVNs)(defaultSlot({})))] : [(0, _vue.h)('div', {
        class: 'vxe-loading--chunk'
      }, [iconSlot || loadingIcon ? (0, _vue.h)('div', {
        class: 'vxe-loading--icon'
      }, iconSlot ? (0, _vn.getSlotVNs)(iconSlot({})) : [(0, _vue.h)('i', {
        class: loadingIcon
      })]) : (0, _vue.h)('div', {
        class: 'vxe-loading--spinner'
      }), textSlot || loadingText ? (0, _vue.h)('div', {
        class: 'vxe-loading--text'
      }, textSlot ? (0, _vn.getSlotVNs)(textSlot({})) : `${loadingText}`) : null])]);
    };
  }
});
