// vxe-pc-ui@4.1.2/es/calendar/src/calendar.js
import { defineComponent, ref, h, reactive } from 'vue';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeCalendar',
    props: {},
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const refElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeMaps = {};
        const $xeCalendar = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderVN = () => {
            const defaultSlot = slots.default;
            return h('div', {
                ref: refElem,
                class: ['vxe-calendar']
            }, defaultSlot ? defaultSlot({}) : []);
        };
        $xeCalendar.renderVN = renderVN;
        return $xeCalendar;
    },
    render() {
        return this.renderVN();
    }
});
