// vxe-pc-ui@4.1.2/lib/select/src/optgroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _util = require("./util");
var _default = exports.default = (0, _vue.defineComponent)({
  name: 'VxeOptgroup',
  props: {
    label: {
      type: [String, Number, Boolean],
      default: ''
    },
    visible: {
      type: Boolean,
      default: null
    },
    className: [String, Function],
    disabled: Boolean
  },
  setup(props, {
    slots
  }) {
    const elem = (0, _vue.ref)();
    const $xeSelect = (0, _vue.inject)('$xeSelect', {});
    const option = (0, _util.createOption)($xeSelect, props);
    const xeOption = {
      option
    };
    option.options = [];
    (0, _vue.provide)('xeoptgroup', xeOption);
    (0, _util.watchOption)(props, option);
    (0, _vue.onMounted)(() => {
      (0, _util.assembleOption)($xeSelect, elem.value, option);
    });
    (0, _vue.onUnmounted)(() => {
      (0, _util.destroyOption)($xeSelect, option);
    });
    return () => {
      return (0, _vue.h)('div', {
        ref: elem
      }, slots.default ? slots.default() : []);
    };
  }
});
