// vxe-pc-ui@4.1.2/es/text/src/text.js
import { defineComponent, ref, h, reactive, createCommentVNode } from 'vue';
import XEUtils from 'xe-utils';
import { VxeUI, getConfig, getIcon, getI18n, useSize } from '../../ui';
import { getSlotVNs } from '../../ui/src/vn';
export default defineComponent({
    name: 'VxeText',
    props: {
        status: String,
        title: [String, Number],
        icon: String,
        content: [String, Number],
        clickToCopy: Boolean,
        size: { type: String, default: () => getConfig().text.size || getConfig().size }
    },
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const { computeSize } = useSize(props);
        const refElem = ref();
        const refContentElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeMaps = {};
        const clickIconEvent = () => {
            const { content, clickToCopy } = props;
            if (clickToCopy) {
                const contentEl = refContentElem.value;
                const copyVal = (contentEl ? contentEl.textContent : '') || content;
                if (copyVal) {
                    if (VxeUI.clipboard.copy(copyVal)) {
                        if (VxeUI.modal) {
                            VxeUI.modal.message({
                                content: getI18n('vxe.text.copySuccess'),
                                status: 'success'
                            });
                        }
                    }
                    else {
                        if (VxeUI.modal) {
                            VxeUI.modal.message({
                                content: getI18n('vxe.text.copyError'),
                                status: 'error'
                            });
                        }
                    }
                }
            }
        };
        const $xeText = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderContent = () => {
            const { icon, content, clickToCopy } = props;
            const defaultSlot = slots.default;
            const iconSlot = slots.icon;
            return [
                iconSlot || icon || clickToCopy
                    ? h('span', {
                        class: 'vxe-text--icon',
                        onClick: clickIconEvent
                    }, iconSlot
                        ? getSlotVNs(iconSlot({}))
                        : [
                            h('i', {
                                class: icon || getIcon().TEXT_COPY
                            })
                        ])
                    : createCommentVNode(),
                h('span', {
                    ref: refContentElem,
                    class: 'vxe-text--content'
                }, defaultSlot ? defaultSlot({}) : XEUtils.toValueString(content))
            ];
        };
        const renderVN = () => {
            const { status, title, clickToCopy } = props;
            const vSize = computeSize.value;
            return h('span', {
                ref: refElem,
                title,
                class: ['vxe-text', {
                        [`size--${vSize}`]: vSize,
                        [`theme--${status}`]: status,
                        'is--copy': clickToCopy
                    }]
            }, renderContent());
        };
        $xeText.renderVN = renderVN;
        return $xeText;
    },
    render() {
        return this.renderVN();
    }
});
