// vxe-pc-ui@4.1.2/lib/ui/src/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.eqEmptyValue = eqEmptyValue;
exports.getFuncText = getFuncText;
exports.getLastZIndex = getLastZIndex;
exports.handleBooleanDefaultValue = handleBooleanDefaultValue;
exports.isEnableConf = isEnableConf;
exports.nextZIndex = nextZIndex;
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _core = require("@vxe-ui/core");
var _domZindex = _interopRequireDefault(require("dom-zindex"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function isEnableConf(conf) {
  return conf && conf.enabled !== false;
}
function nextZIndex() {
  return _domZindex.default.getNext();
}
function getLastZIndex() {
  return _domZindex.default.getCurrent();
}
function getFuncText(content, args) {
  if (content) {
    const translate = (0, _core.getConfig)().translate;
    return _xeUtils.default.toValueString(translate ? translate('' + content, args) : content);
  }
  return '';
}
/**
 * 判断值为：'' | null | undefined 时都属于空值
 */
function eqEmptyValue(cellValue) {
  return cellValue === null || cellValue === undefined || cellValue === '';
}
function handleBooleanDefaultValue(value) {
  return _xeUtils.default.isBoolean(value) ? value : null;
}
