// vxe-pc-ui@4.1.2/es/layout-header/src/layout-header.js
import { defineComponent, ref, h, reactive } from 'vue';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeLayoutHeader',
    props: {
        fixed: Boolean
    },
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const refElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeMaps = {};
        const $xeLayoutHeader = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderVN = () => {
            const defaultSlot = slots.default;
            return h('header', {
                ref: refElem,
                class: ['vxe-layout-header', {
                        'is--fixed': props.fixed
                    }]
            }, defaultSlot ? defaultSlot({}) : []);
        };
        $xeLayoutHeader.renderVN = renderVN;
        return $xeLayoutHeader;
    },
    render() {
        return this.renderVN();
    }
});
