// vxe-pc-ui@4.1.2/lib/print/src/page-break.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _util = require("./util");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = exports.default = (0, _vue.defineComponent)({
  name: 'VxePrintPageBreak',
  props: {},
  emits: [],
  setup(props, context) {
    const {
      slots
    } = context;
    const xID = _xeUtils.default.uniqueId();
    const $xePrint = (0, _vue.inject)('$xePrint', null);
    const refElem = (0, _vue.ref)();
    const reactData = (0, _vue.reactive)({});
    const refMaps = {
      refElem
    };
    const computeMaps = {};
    const pageBreakConfig = (0, _vue.reactive)({
      id: xID,
      slots
    });
    const $xePrintPageBreak = {
      xID,
      props,
      context,
      reactData,
      getRefMaps: () => refMaps,
      getComputeMaps: () => computeMaps
    };
    if (!$xePrint) {
      $xePrintPageBreak.renderVN = () => [];
      return $xePrintPageBreak;
    }
    const renderVN = () => {
      return (0, _vue.h)('div', {
        ref: refElem
      });
    };
    $xePrintPageBreak.renderVN = renderVN;
    (0, _vue.onMounted)(() => {
      if ($xePrint && refElem.value) {
        (0, _util.assemblePageBreak)($xePrint, refElem.value, pageBreakConfig);
      }
    });
    (0, _vue.onUnmounted)(() => {
      if ($xePrint) {
        (0, _util.destroyPageBreak)($xePrint, pageBreakConfig);
      }
    });
    return $xePrintPageBreak;
  },
  render() {
    return this.renderVN();
  }
});
