// vxe-pc-ui@4.1.2/lib/tabs/src/tab-pane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _util = require("./util");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = exports.default = (0, _vue.defineComponent)({
  name: 'VxeTabPane',
  props: {
    title: [String, Number],
    name: [String, Number],
    icon: String,
    titleWidth: [String, Number],
    titleAlign: [String, Number],
    preload: Boolean,
    permissionCode: [String, Number]
  },
  emits: [],
  setup(props, context) {
    const {
      slots
    } = context;
    const xID = _xeUtils.default.uniqueId();
    const refElem = (0, _vue.ref)();
    const $xeTabs = (0, _vue.inject)('$xeTabs', null);
    const reactData = (0, _vue.reactive)({});
    const tabConfig = (0, _vue.reactive)({
      id: xID,
      title: props.title,
      name: props.name,
      icon: props.icon,
      titleWidth: props.titleWidth,
      titleAlign: props.titleAlign,
      preload: props.preload,
      permissionCode: props.permissionCode,
      slots: {
        default: slots.default
      }
    });
    const refMaps = {
      refElem
    };
    const computeMaps = {};
    const $xeTabPane = {
      xID,
      props,
      context,
      reactData,
      getRefMaps: () => refMaps,
      getComputeMaps: () => computeMaps
    };
    (0, _vue.watch)(() => props.title, val => {
      tabConfig.title = val;
    });
    (0, _vue.watch)(() => props.name, val => {
      tabConfig.name = val;
    });
    (0, _vue.watch)(() => props.icon, val => {
      tabConfig.icon = val;
    });
    (0, _vue.watch)(() => props.permissionCode, val => {
      tabConfig.permissionCode = val;
    });
    (0, _vue.onMounted)(() => {
      if ($xeTabs && refElem.value) {
        (0, _util.assembleAnchorTab)($xeTabs, refElem.value, tabConfig);
      }
    });
    (0, _vue.onUnmounted)(() => {
      if ($xeTabs) {
        (0, _util.destroyAnchorTab)($xeTabs, tabConfig);
      }
    });
    const renderVN = () => {
      return (0, _vue.h)('div', {
        ref: refElem
      }, []);
    };
    $xeTabPane.renderVN = renderVN;
    return $xeTabPane;
  },
  render() {
    return this.renderVN();
  }
});
