// @vxe-ui/core@4.0.3/lib/src/core.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  setTheme: true,
  getTheme: true,
  setLanguage: true,
  setI18n: true,
  setConfig: true,
  getConfig: true,
  setIcon: true,
  getIcon: true,
  coreVersion: true,
  use: true,
  getComponent: true,
  component: true,
  VxeUI: true
};
exports.VxeUI = void 0;
exports.component = component;
exports.default = exports.coreVersion = void 0;
exports.getComponent = getComponent;
exports.getConfig = getConfig;
exports.getIcon = getIcon;
exports.getTheme = getTheme;
exports.setConfig = setConfig;
exports.setI18n = setI18n;
exports.setIcon = setIcon;
exports.setLanguage = setLanguage;
exports.setTheme = setTheme;
exports.use = use;
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _domZindex = _interopRequireDefault(require("dom-zindex"));
var _globalStore = require("./globalStore");
var _iconStore = require("./iconStore");
var _themeStore = require("./themeStore");
var _i18nStore = require("./i18nStore");
var _event = require("./event");
Object.keys(_event).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _event[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _event[key];
    }
  });
});
var _resize = require("./resize");
Object.keys(_resize).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _resize[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _resize[key];
    }
  });
});
var _i18n = require("./i18n");
Object.keys(_i18n).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _i18n[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _i18n[key];
    }
  });
});
var _renderer = require("./renderer");
Object.keys(_renderer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _renderer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _renderer[key];
    }
  });
});
var _validators = require("./validators");
Object.keys(_validators).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _validators[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _validators[key];
    }
  });
});
var _menus = require("./menus");
Object.keys(_menus).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menus[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _menus[key];
    }
  });
});
var _formats = require("./formats");
Object.keys(_formats).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formats[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _formats[key];
    }
  });
});
var _commands = require("./commands");
Object.keys(_commands).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _commands[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _commands[key];
    }
  });
});
var _interceptor = require("./interceptor");
Object.keys(_interceptor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interceptor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interceptor[key];
    }
  });
});
var _clipboard = require("./clipboard");
Object.keys(_clipboard).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _clipboard[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _clipboard[key];
    }
  });
});
var _permission = require("./permission");
Object.keys(_permission).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _permission[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _permission[key];
    }
  });
});
var _log = require("./log");
Object.keys(_log).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _log[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _log[key];
    }
  });
});
var _hooks = require("./hooks");
Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});
var _useFns = require("./useFns");
Object.keys(_useFns).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useFns[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useFns[key];
    }
  });
});
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function setTheme(name) {
  const theme = !name || name === 'default' ? 'light' : name;
  _themeStore.themeConfigStore.theme = theme;
  if (typeof document !== 'undefined') {
    const documentElement = document.documentElement;
    if (documentElement) {
      documentElement.setAttribute('data-vxe-ui-theme', theme);
    }
  }
  return VxeUI;
}
function getTheme() {
  return _themeStore.themeConfigStore.theme;
}
function setLanguage(locale) {
  _i18nStore.i18nConfigStore.language = locale || 'zh-CN';
  return VxeUI;
}
function setI18n(locale, data) {
  _i18nStore.i18nConfigStore.langMaps[locale] = Object.assign({}, data);
  return VxeUI;
}
/**
* 全局参数设置
*/
function setConfig(options) {
  if (options) {
    if (options.zIndex) {
      _domZindex.default.setCurrent(options.zIndex);
    }
    if (options.theme) {
      setTheme(options.theme);
    }
    _xeUtils.default.merge(_globalStore.globalConfigStore, options);
  }
  return VxeUI;
}
/**
* 获取全局参数
*/
function getConfig(key, defaultValue) {
  return arguments.length ? _xeUtils.default.get(_globalStore.globalConfigStore, key, defaultValue) : _globalStore.globalConfigStore;
}
function setIcon(options) {
  if (options) {
    Object.assign(_iconStore.iconConfigStore, options);
  }
  return VxeUI;
}
function getIcon(key) {
  return arguments.length ? _xeUtils.default.get(_iconStore.iconConfigStore, key) : _iconStore.iconConfigStore;
}
const coreVersion = exports.coreVersion = "4.0.3";
const installedPlugins = [];
function use(Plugin, options) {
  if (Plugin && Plugin.install) {
    if (installedPlugins.indexOf(Plugin) === -1) {
      Plugin.install(VxeUI, options);
      installedPlugins.push(Plugin);
    }
  }
  return VxeUI;
}
const components = {};
function getComponent(name) {
  return components[name] || null;
}
function component(comp) {
  if (comp && comp.name) {
    components[comp.name] = comp;
  }
}
const VxeUI = exports.VxeUI = {
  coreVersion,
  setTheme,
  getTheme,
  setConfig,
  getConfig: getConfig,
  setIcon,
  getIcon: getIcon,
  setLanguage,
  hasLanguage: _i18n.hasLanguage,
  getLanguage: _i18n.getLanguage,
  setI18n,
  getI18n: _i18n.getI18n,
  globalEvents: _event.globalEvents,
  GLOBAL_EVENT_KEYS: _event.GLOBAL_EVENT_KEYS,
  createEvent: _event.createEvent,
  globalResize: _resize.globalResize,
  renderer: _renderer.renderer,
  validators: _validators.validators,
  menus: _menus.menus,
  formats: _formats.formats,
  commands: _commands.commands,
  interceptor: _interceptor.interceptor,
  clipboard: _clipboard.clipboard,
  log: _log.log,
  permission: _permission.permission,
  hooks: _hooks.hooks,
  component,
  getComponent,
  useFns: _useFns.useFns,
  use
};
setTheme();
var _default = exports.default = VxeUI;
