// vxe-pc-ui@4.1.2/es/flow-design/src/flow-design.js
import { defineComponent, ref, h, reactive } from 'vue';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeFlowDesign',
    props: {},
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const refElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeMaps = {};
        const $xeFlowDesign = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderVN = () => {
            const defaultSlot = slots.default;
            return h('div', {
                ref: refElem,
                class: ['vxe-upload']
            }, defaultSlot ? defaultSlot({}) : []);
        };
        $xeFlowDesign.renderVN = renderVN;
        return $xeFlowDesign;
    },
    render() {
        return this.renderVN();
    }
});
