// vxe-pc-ui@4.1.2/lib/list-design/src/layout-setting.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _core = require("@vxe-ui/core");
var _tabs = _interopRequireDefault(require("../../tabs/src/tabs"));
var _tabPane = _interopRequireDefault(require("../../tabs/src/tab-pane"));
var _defaultSettingForm = require("./default-setting-form");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = exports.default = (0, _vue.defineComponent)({
  name: 'ListDesignLayoutSetting',
  props: {},
  emits: [],
  setup() {
    const activeTab = (0, _vue.ref)(1);
    return () => {
      return (0, _vue.h)('div', {
        class: 'vxe-list-design--setting'
      }, [(0, _vue.h)('div', {
        class: 'vxe-list-design--setting-form'
      }, [(0, _vue.h)(_tabs.default, {
        modelValue: activeTab.value,
        titleWidth: '50%',
        titleAlign: 'center',
        padding: true,
        class: 'vxe-list-design--setting-form-tabs',
        'onUpdate:modelValue'(val) {
          activeTab.value = val;
        }
      }, {
        default() {
          return [(0, _vue.h)(_tabPane.default, {
            title: (0, _core.getI18n)('vxe.listDesign.fieldSettingTab'),
            icon: (0, _core.getIcon)().LIST_DESIGN_FIELD_SETTING,
            name: 1
          }, {
            default() {
              return (0, _vue.h)(_defaultSettingForm.DefaultFieldSettingFormComponent);
            }
          }), (0, _vue.h)(_tabPane.default, {
            title: (0, _core.getI18n)('vxe.listDesign.listSettingTab'),
            icon: (0, _core.getIcon)().LIST_DESIGN_LIST_SETTING,
            name: 2
          }, {
            default() {
              return (0, _vue.h)(_defaultSettingForm.DefaultParameterSettingFormComponent);
            }
          })];
        }
      })])]);
    };
  }
});
