// vxe-pc-ui@4.1.2/lib/button/src/button-group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _core = require("@vxe-ui/core");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _button = _interopRequireDefault(require("../../button/src/button"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = exports.default = (0, _vue.defineComponent)({
  name: 'VxeButtonGroup',
  props: {
    options: Array,
    mode: String,
    status: String,
    round: Boolean,
    circle: Boolean,
    className: [String, Function],
    disabled: Boolean,
    permissionCode: [String, Number],
    size: {
      type: String,
      default: () => (0, _core.getConfig)().buttonGroup.size || (0, _core.getConfig)().size
    }
  },
  emits: ['click'],
  setup(props, context) {
    const {
      slots,
      emit
    } = context;
    const xID = _xeUtils.default.uniqueId();
    const computeMaps = {};
    const $xeButtonGroup = {
      xID,
      props,
      context,
      getComputeMaps: () => computeMaps
    };
    (0, _core.useSize)(props);
    const {
      computePermissionInfo
    } = (0, _core.usePermission)(props);
    const buttonGroupMethods = {
      dispatchEvent(type, params, evnt) {
        emit(type, (0, _core.createEvent)(evnt, {
          $buttonGroup: $xeButtonGroup
        }, params));
      }
    };
    const buttonGroupPrivateMethods = {
      handleClick(params, evnt) {
        const {
          options
        } = props;
        const {
          name
        } = params;
        const option = options ? options.find(item => item.name === name) : null;
        buttonGroupMethods.dispatchEvent('click', Object.assign(Object.assign({}, params), {
          option
        }), evnt);
      }
    };
    Object.assign($xeButtonGroup, buttonGroupMethods, buttonGroupPrivateMethods);
    const renderVN = () => {
      const {
        className,
        options
      } = props;
      const permissionInfo = computePermissionInfo.value;
      const defaultSlot = slots.default;
      if (!permissionInfo.visible) {
        return (0, _vue.createCommentVNode)();
      }
      return (0, _vue.h)('div', {
        class: ['vxe-button-group', className ? _xeUtils.default.isFunction(className) ? className({
          $buttonGroup: $xeButtonGroup
        }) : className : '']
      }, defaultSlot ? defaultSlot({}) : options ? options.map((item, index) => {
        return (0, _vue.h)(_button.default, Object.assign({
          key: index
        }, item));
      }) : []);
    };
    $xeButtonGroup.renderVN = renderVN;
    (0, _vue.provide)('$xeButtonGroup', $xeButtonGroup);
    return renderVN;
  }
});
