// vxe-pc-ui@4.1.2/es/collapse-pane/src/collapse-pane.js
import { defineComponent, ref, h, reactive } from 'vue';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeCollapsePane',
    props: {},
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const refElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeMaps = {};
        const $xeCollapsePane = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderVN = () => {
            const defaultSlot = slots.default;
            return h('div', {
                ref: refElem,
                class: ['vxe-collapse-pane']
            }, defaultSlot ? defaultSlot({}) : []);
        };
        $xeCollapsePane.renderVN = renderVN;
        return $xeCollapsePane;
    },
    render() {
        return this.renderVN();
    }
});
