// vxe-pc-ui@4.1.2/es/form/src/form-gather.js
import { defineComponent, h, onUnmounted, inject, ref, reactive, onMounted, provide } from 'vue';
import { createItem, watchItem, destroyItem, assembleItem } from './util';
import { formItemProps } from './form-item';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeFormGather',
    props: formItemProps,
    setup(props, { slots }) {
        const refElem = ref();
        const $xeForm = inject('$xeForm', {});
        const parentFormGather = inject('$xeFormGather', null);
        const formItem = reactive(createItem($xeForm, props));
        formItem.children = [];
        const formItemInfo = { itemConfig: formItem };
        provide('xeFormItemInfo', formItemInfo);
        watchItem(props, formItem);
        onMounted(() => {
            assembleItem($xeForm, refElem.value, formItem, parentFormGather);
        });
        onUnmounted(() => {
            destroyItem($xeForm, formItem);
        });
        const renderVN = () => {
            const { className, field } = props;
            const span = props.span || ($xeForm ? $xeForm.props.span : null);
            const defaultSlot = slots.default;
            return h('div', {
                ref: refElem,
                class: ['vxe-form--gather vxe-form--item-row', formItem.id, span ? `vxe-form--item-col_${span} is--span` : '', className ? (XEUtils.isFunction(className) ? className({ $form: $xeForm, data: $xeForm ? $xeForm.props.data : {}, item: formItem, field: field, property: field }) : className) : '']
            }, defaultSlot ? defaultSlot({}) : []);
        };
        const $xeFormGather = {
            formItem,
            renderVN
        };
        provide('$xeFormGather', $xeFormGather);
        provide('$xeFormItem', null);
        return $xeFormGather;
    },
    render() {
        return this.renderVN();
    }
});
