// vxe-pc-ui@4.1.2/es/layout-aside/src/layout-aside.js
import { defineComponent, ref, h, reactive, onMounted, computed, provide } from 'vue';
import { toCssUnit } from '../../ui/src/dom';
import VxeLoadingComponent from '../../loading/src/loading';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeLayoutAside',
    props: {
        width: [String, Number],
        collapsed: Boolean,
        collapseWidth: [String, Number],
        loading: Boolean,
        padding: Boolean
    },
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const refElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeWrapperWidth = computed(() => {
            const { width, collapsed, collapseWidth } = props;
            if (collapsed) {
                if (collapseWidth) {
                    return toCssUnit(collapseWidth);
                }
            }
            else {
                if (width) {
                    return toCssUnit(width);
                }
            }
            return '';
        });
        const computeMaps = {};
        const $xeLayoutAside = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderVN = () => {
            const { width, collapsed, loading, padding } = props;
            const wrapperWidth = computeWrapperWidth.value;
            const defaultSlot = slots.default;
            return h('aside', {
                ref: refElem,
                class: ['vxe-layout-aside', {
                        'is--padding': padding,
                        'is--default-width': !width,
                        'is--collapse': collapsed,
                        'is--loading': loading
                    }],
                style: wrapperWidth
                    ? {
                        width: wrapperWidth
                    }
                    : null
            }, [
                h('div', {
                    class: 'vxe-layout-aside--inner'
                }, defaultSlot ? defaultSlot({}) : []),
                /**
                 * 加载中
                 */
                h(VxeLoadingComponent, {
                    class: 'vxe-list-view--loading',
                    modelValue: loading
                })
            ]);
        };
        onMounted(() => {
        });
        $xeLayoutAside.renderVN = renderVN;
        provide('$xeLayoutAside', $xeLayoutAside);
        return $xeLayoutAside;
    },
    render() {
        return this.renderVN();
    }
});
