// vxe-table@4.6.18/lib/v-x-e-table/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  use: true,
  t: true,
  _t: true,
  setConfig: true,
  globalConfs: true,
  v: true,
  setup: true,
  config: true,
  setIcon: true,
  globalStore: true,
  getComponent: true,
  component: true,
  version: true,
  tableVersion: true,
  VXETable: true,
  VxeUI: true
};
exports.VxeUI = exports.VXETable = void 0;
exports._t = _t;
exports.component = component;
exports.default = exports.config = void 0;
exports.getComponent = getComponent;
exports.setConfig = exports.globalStore = exports.globalConfs = void 0;
exports.setIcon = setIcon;
exports.setup = void 0;
exports.t = t;
exports.tableVersion = void 0;
exports.use = use;
exports.version = exports.v = void 0;
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _conf = _interopRequireDefault(require("./src/conf"));
var _domZindex = _interopRequireDefault(require("dom-zindex"));
var _interceptor = require("./src/interceptor");
Object.keys(_interceptor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interceptor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interceptor[key];
    }
  });
});
var _renderer = require("./src/renderer");
Object.keys(_renderer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _renderer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _renderer[key];
    }
  });
});
var _commands = require("./src/commands");
Object.keys(_commands).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _commands[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _commands[key];
    }
  });
});
var _menus = require("./src/menus");
Object.keys(_menus).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menus[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _menus[key];
    }
  });
});
var _formats = require("./src/formats");
Object.keys(_formats).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formats[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _formats[key];
    }
  });
});
var _validators = require("./src/validators");
Object.keys(_validators).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _validators[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _validators[key];
    }
  });
});
var _hooks = require("./src/hooks");
Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});
var _theme = require("./src/theme");
var _utils = require("../tools/utils");
var _log = require("../tools/log");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function getExportOrImpotType(types, flag) {
  const rest = [];
  _xeUtils.default.objectEach(types, (val, type) => {
    if (val === 0 || val === flag) {
      rest.push(type);
    }
  });
  return rest;
}
const installedPlugins = [];
function use(Plugin, options) {
  /* eslint-disable @typescript-eslint/no-use-before-define */
  if (Plugin && Plugin.install) {
    if (installedPlugins.indexOf(Plugin) === -1) {
      Plugin.install(VXETable, options);
      installedPlugins.push(Plugin);
    }
  }
  return VXETable;
}
function t(key, args) {
  return _conf.default.i18n(key, args);
}
function _t(key, args) {
  return key ? _xeUtils.default.toValueString(_conf.default.translate ? _conf.default.translate(key, args) : key) : '';
}
/**
 * 全局参数设置
 */
const setConfig = options => {
  if (options) {
    if (options.theme) {
      (0, _theme.setTheme)(options.theme);
    }
    if (options.zIndex) {
      _domZindex.default.setCurrent(options.zIndex);
    }
    _xeUtils.default.merge(_conf.default, options);
  }
  return VXETable;
};
exports.setConfig = setConfig;
class VXETableConfig {
  /**
   * 获取当前的 zIndex
   */
  get zIndex() {
    return (0, _utils.getLastZIndex)();
  }
  /**
   * 获取下一个 zIndex
   */
  get nextZIndex() {
    return (0, _utils.nextZIndex)();
  }
  /**
   * 获取所有导出类型
   */
  get exportTypes() {
    return getExportOrImpotType(_conf.default.export.types, 1);
  }
  /**
   * 获取所有导入类型
   */
  get importTypes() {
    return getExportOrImpotType(_conf.default.export.types, 2);
  }
}
// 已废弃
const globalConfs = exports.globalConfs = new VXETableConfig();
const v = exports.v = 'v4';
/**
 * 已废弃，请使用 setConfig
 * @deprecated
 */
const setup = options => {
  if (process.env.NODE_ENV === 'development') {
    (0, _log.warnLog)('vxe.error.delFunc', ['setup', 'setConfig']);
  }
  setConfig(options);
  return _conf.default;
};
/**
 * 已废弃，请使用 setConfig
 * @deprecated
 */
exports.setup = setup;
const config = options => {
  if (process.env.NODE_ENV === 'development') {
    (0, _log.warnLog)('vxe.error.delFunc', ['setup', 'setConfig']);
  }
  setConfig(options);
  return _conf.default;
};
exports.config = config;
function setIcon(options) {
  if (options) {
    Object.assign(_conf.default.icon, options);
  }
  return VXETable;
}
const globalStore = exports.globalStore = {};
const components = {};
function getComponent(name) {
  return components[name] || null;
}
function component(comp) {
  if (comp && comp.name) {
    components[comp.name] = comp;
  }
}
const version = exports.version = "4.6.18";
const tableVersion = exports.tableVersion = version;
const VXETable = exports.VXETable = {
  v,
  version,
  tableVersion,
  setConfig,
  setIcon,
  globalStore,
  interceptor: _interceptor.interceptor,
  renderer: _renderer.renderer,
  commands: _commands.commands,
  formats: _formats.formats,
  validators: _validators.validators,
  menus: _menus.menus,
  hooks: _hooks.hooks,
  use,
  t,
  _t,
  setTheme: _theme.setTheme,
  getTheme: _theme.getTheme,
  getComponent,
  // 已废弃
  config,
  setup,
  globalConfs
};
const VxeUI = exports.VxeUI = VXETable;
(0, _theme.setTheme)('light');
var _default = exports.default = VXETable;
