// vxe-pc-ui@4.1.2/lib/form-design/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.VxeFormDesign = exports.FormDesign = void 0;
var _core = require("@vxe-ui/core");
var _formDesign = _interopRequireDefault(require("./src/form-design"));
var _use = require("./src/use");
var _dynamics = require("../dynamics");
require("./render");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const VxeFormDesign = exports.VxeFormDesign = Object.assign({}, _formDesign.default, {
  install(app) {
    app.component(_formDesign.default.name, _formDesign.default);
  }
});
const formDesign = {
  useWidgetName: _use.useWidgetName,
  useWidgetView: _use.useWidgetView,
  useWidgetPropDataSource: _use.useWidgetPropDataSource
};
_dynamics.dynamicApp.component(_formDesign.default.name, _formDesign.default);
_core.VxeUI.component(_formDesign.default);
_core.VxeUI.formDesign = formDesign;
const FormDesign = exports.FormDesign = VxeFormDesign;
var _default = exports.default = VxeFormDesign;
