// vxe-pc-ui@4.1.2/es/layout-footer/src/layout-footer.js
import { defineComponent, ref, h, reactive } from 'vue';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeLayoutFooter',
    props: {
        fixed: Boolean,
        align: String
    },
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const refElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeMaps = {};
        const $xeLayoutFooter = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderVN = () => {
            const { fixed, align } = props;
            const defaultSlot = slots.default;
            return h('footer', {
                ref: refElem,
                class: ['vxe-layout-footer', align ? `align--${align}` : '', {
                        'is--fixed': fixed
                    }]
            }, defaultSlot ? defaultSlot({}) : []);
        };
        $xeLayoutFooter.renderVN = renderVN;
        return $xeLayoutFooter;
    },
    render() {
        return this.renderVN();
    }
});
