// vxe-pc-ui@4.1.2/lib/form/src/form-gather.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _vue = require("vue");
var _util = require("./util");
var _formItem = require("./form-item");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = exports.default = (0, _vue.defineComponent)({
  name: 'VxeFormGather',
  props: _formItem.formItemProps,
  setup(props, {
    slots
  }) {
    const refElem = (0, _vue.ref)();
    const $xeForm = (0, _vue.inject)('$xeForm', {});
    const parentFormGather = (0, _vue.inject)('$xeFormGather', null);
    const formItem = (0, _vue.reactive)((0, _util.createItem)($xeForm, props));
    formItem.children = [];
    const formItemInfo = {
      itemConfig: formItem
    };
    (0, _vue.provide)('xeFormItemInfo', formItemInfo);
    (0, _util.watchItem)(props, formItem);
    (0, _vue.onMounted)(() => {
      (0, _util.assembleItem)($xeForm, refElem.value, formItem, parentFormGather);
    });
    (0, _vue.onUnmounted)(() => {
      (0, _util.destroyItem)($xeForm, formItem);
    });
    const renderVN = () => {
      const {
        className,
        field
      } = props;
      const span = props.span || ($xeForm ? $xeForm.props.span : null);
      const defaultSlot = slots.default;
      return (0, _vue.h)('div', {
        ref: refElem,
        class: ['vxe-form--gather vxe-form--item-row', formItem.id, span ? `vxe-form--item-col_${span} is--span` : '', className ? _xeUtils.default.isFunction(className) ? className({
          $form: $xeForm,
          data: $xeForm ? $xeForm.props.data : {},
          item: formItem,
          field: field,
          property: field
        }) : className : '']
      }, defaultSlot ? defaultSlot({}) : []);
    };
    const $xeFormGather = {
      formItem,
      renderVN
    };
    (0, _vue.provide)('$xeFormGather', $xeFormGather);
    (0, _vue.provide)('$xeFormItem', null);
    return $xeFormGather;
  },
  render() {
    return this.renderVN();
  }
});
