// vxe-pc-ui@4.1.2/es/select/src/optgroup.js
import { defineComponent, h, onUnmounted, provide, inject, ref, onMounted } from 'vue';
import { createOption, watchOption, destroyOption, assembleOption } from './util';
export default defineComponent({
    name: 'VxeOptgroup',
    props: {
        label: { type: [String, Number, Boolean], default: '' },
        visible: { type: Boolean, default: null },
        className: [String, Function],
        disabled: Boolean
    },
    setup(props, { slots }) {
        const elem = ref();
        const $xeSelect = inject('$xeSelect', {});
        const option = createOption($xeSelect, props);
        const xeOption = { option };
        option.options = [];
        provide('xeoptgroup', xeOption);
        watchOption(props, option);
        onMounted(() => {
            assembleOption($xeSelect, elem.value, option);
        });
        onUnmounted(() => {
            destroyOption($xeSelect, option);
        });
        return () => {
            return h('div', {
                ref: elem
            }, slots.default ? slots.default() : []);
        };
    }
});
