// vxe-pc-ui@4.1.2/es/select/src/option.js
import { defineComponent, h, onUnmounted, inject, ref, onMounted } from 'vue';
import { createOption, watchOption, destroyOption, assembleOption } from './util';
export default defineComponent({
    name: 'VxeOption',
    props: {
        value: null,
        label: { type: [String, Number, Boolean], default: '' },
        visible: { type: Boolean, default: null },
        className: [String, Function],
        disabled: Boolean
    },
    setup(props, { slots }) {
        const elem = ref();
        const $xeSelect = inject('$xeSelect', {});
        const optGroup = inject('xeoptgroup', null);
        const option = createOption($xeSelect, props);
        option.slots = slots;
        watchOption(props, option);
        onMounted(() => {
            assembleOption($xeSelect, elem.value, option, optGroup);
        });
        onUnmounted(() => {
            destroyOption($xeSelect, option);
        });
        return () => {
            return h('div', {
                ref: elem
            });
        };
    }
});
