// vxe-pc-ui@4.1.2/lib/form-design/src/widget-info.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormDesignWidgetInfo = void 0;
exports.configToWidget = configToWidget;
exports.createHandleGetField = void 0;
exports.getWidgetConfigCustomGroup = getWidgetConfigCustomGroup;
exports.getWidgetConfigScope = exports.getWidgetConfigIcon = exports.getWidgetConfigGroup = void 0;
exports.getWidgetConfigTitle = getWidgetConfigTitle;
exports.getWidgetConfigUnique = void 0;
var _vue = require("vue");
var _core = require("@vxe-ui/core");
var _util = require("./util");
var _log = require("../../ui/src/log");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
// 控件原始配置信息，带响应
const refWidgetReactConfigMaps = (0, _vue.ref)({});
const createHandleGetField = key => {
  return function (name) {
    const widgetReactConfigMaps = refWidgetReactConfigMaps.value;
    const configMaps = widgetReactConfigMaps[name];
    return configMaps ? configMaps[key] : null;
  };
};
exports.createHandleGetField = createHandleGetField;
function getWidgetConfigTitle(name, $xeFormDesign) {
  const widgetReactConfigMaps = refWidgetReactConfigMaps.value;
  const configMaps = widgetReactConfigMaps[name];
  if (configMaps) {
    const configTitle = configMaps.title;
    const params = {
      name,
      $formDesign: $xeFormDesign
    };
    return _xeUtils.default.toValueString(_xeUtils.default.isFunction(configTitle) ? configTitle(params) : configTitle);
  }
  return name;
}
function getWidgetConfigCustomGroup(name, $xeFormDesign) {
  const widgetReactConfigMaps = refWidgetReactConfigMaps.value;
  const configMaps = widgetReactConfigMaps[name];
  if (configMaps) {
    const configCustomGroup = configMaps.customGroup;
    const params = {
      name,
      $formDesign: $xeFormDesign
    };
    return _xeUtils.default.toValueString(_xeUtils.default.isFunction(configCustomGroup) ? configCustomGroup(params) : configCustomGroup);
  }
  return name;
}
const getWidgetConfigIcon = exports.getWidgetConfigIcon = createHandleGetField('icon');
const getWidgetConfigGroup = exports.getWidgetConfigGroup = createHandleGetField('group');
const getWidgetConfigUnique = exports.getWidgetConfigUnique = createHandleGetField('unique');
const getWidgetConfigScope = exports.getWidgetConfigScope = createHandleGetField('scope');
const validWidgetUniqueField = (field, widgetObjList) => {
  return !_xeUtils.default.findTree(widgetObjList, item => item.field === field, {
    children: 'children'
  });
};
class FormDesignWidgetInfo {
  constructor($xeFormDesign, name, widgetObjList) {
    Object.defineProperty(this, "id", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: 0
    });
    Object.defineProperty(this, "field", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: ''
    });
    Object.defineProperty(this, "title", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: ''
    });
    Object.defineProperty(this, "name", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: ''
    });
    Object.defineProperty(this, "required", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: false
    });
    Object.defineProperty(this, "hidden", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: false
    });
    Object.defineProperty(this, "options", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: {}
    });
    Object.defineProperty(this, "children", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: []
    });
    Object.defineProperty(this, "model", {
      enumerable: true,
      configurable: true,
      writable: true,
      value: {
        update: false,
        value: ''
      }
    });
    let customField = '';
    if (name) {
      const compConf = _core.renderer.get(name) || {};
      if (compConf) {
        const widgetReactConfigMaps = refWidgetReactConfigMaps.value;
        const createWidgetFormConfig = compConf.createFormDesignWidgetConfig;
        if (createWidgetFormConfig) {
          const params = {
            name,
            $formDesign: $xeFormDesign
          };
          const widgetConfig = createWidgetFormConfig(params) || {};
          const titleConf = widgetConfig.title;
          const fieldConf = widgetConfig.field;
          this.title = _xeUtils.default.toValueString(_xeUtils.default.isFunction(titleConf) ? titleConf(params) : titleConf);
          this.options = widgetConfig.options || {};
          this.children = widgetConfig.children || [];
          if (fieldConf) {
            if (_xeUtils.default.isFunction(fieldConf)) {
              customField = fieldConf({
                name,
                $formDesign: $xeFormDesign
              });
            } else {
              customField = fieldConf;
            }
          }
          if (!widgetReactConfigMaps[name]) {
            widgetReactConfigMaps[name] = {
              title: titleConf || '',
              field: fieldConf || '',
              icon: widgetConfig.icon || '',
              group: widgetConfig.group || '',
              unique: widgetConfig.unique || false,
              scope: widgetConfig.scope || '',
              customGroup: widgetConfig.customGroup || ''
            };
            refWidgetReactConfigMaps.value = Object.assign({}, widgetReactConfigMaps);
          }
        }
      }
    }
    const widgetId = (0, _util.getNewWidgetId)(widgetObjList);
    if (customField) {
      // 如果使用了自定义字段，验证字段名是否唯一
      if (!validWidgetUniqueField(customField, widgetObjList)) {
        (0, _log.errLog)('vxe.error.uniField', [customField]);
      }
    }
    this.id = widgetId;
    this.field = customField || `${name}${widgetId}`;
    this.name = name;
  }
}
exports.FormDesignWidgetInfo = FormDesignWidgetInfo;
function configToWidget(conf) {
  const widget = new FormDesignWidgetInfo(null, conf.name, []);
  widget.id = conf.id;
  widget.title = conf.title || '';
  widget.field = conf.field || '';
  widget.required = conf.required || false;
  widget.hidden = conf.hidden || false;
  widget.options = Object.assign({}, widget.options, conf.options);
  widget.children = conf.children ? conf.children.map(item => configToWidget(item)) : [];
  return widget;
}
