// vxe-pc-ui@4.1.2/es/layout-container/src/layout-container.js
import { defineComponent, ref, h, reactive } from 'vue';
import XEUtils from 'xe-utils';
export default defineComponent({
    name: 'VxeLayoutContainer',
    props: {
        vertical: Boolean
    },
    emits: [],
    setup(props, context) {
        const { slots } = context;
        const xID = XEUtils.uniqueId();
        const refElem = ref();
        const reactData = reactive({});
        const refMaps = {
            refElem
        };
        const computeMaps = {};
        const $xeLayoutContainer = {
            xID,
            props,
            context,
            reactData,
            getRefMaps: () => refMaps,
            getComputeMaps: () => computeMaps
        };
        const renderVN = () => {
            const { vertical } = props;
            const defaultSlot = slots.default;
            return h('div', {
                ref: refElem,
                class: ['vxe-layout-container', {
                        'is--vertical': vertical
                    }]
            }, defaultSlot ? defaultSlot({}) : []);
        };
        $xeLayoutContainer.renderVN = renderVN;
        return $xeLayoutContainer;
    },
    render() {
        return this.renderVN();
    }
});
