// vxe-pc-ui@4.1.2/lib/form-design/widget-vxe-upload-image/vxe-upload-image-form.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WidgetVxeUploadImageFormComponent = void 0;
var _vue = require("vue");
var _core = require("@vxe-ui/core");
var _vxeUploadFileData = require("../widget-vxe-upload-file/vxe-upload-file-data");
var _use = require("../../form-design/src/use");
var _form = _interopRequireDefault(require("../../form/src/form"));
var _formItem = _interopRequireDefault(require("../../form/src/form-item"));
var _input = _interopRequireDefault(require("../../input/src/input"));
var _switch = _interopRequireDefault(require("../../switch/src/switch"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const WidgetVxeUploadImageFormComponent = exports.WidgetVxeUploadImageFormComponent = (0, _vue.defineComponent)({
  props: {
    renderOpts: {
      type: Object,
      default: () => ({})
    },
    renderParams: {
      type: Object,
      default: () => ({})
    }
  },
  emits: [],
  setup(props) {
    const {
      computeKebabCaseName
    } = (0, _use.useWidgetName)(props);
    const limitSizeOptions = (0, _vue.ref)((0, _vxeUploadFileData.getLimitSizeOptions)());
    return () => {
      const {
        renderParams
      } = props;
      const {
        widget
      } = renderParams;
      const {
        options
      } = widget;
      const kebabCaseName = computeKebabCaseName.value;
      return (0, _vue.h)(_form.default, {
        class: ['vxe-form-design--widget-render-form-wrapper', `widget-${kebabCaseName}`],
        vertical: true,
        span: 24,
        titleBold: true,
        titleOverflow: true,
        data: options
      }, {
        default() {
          return [(0, _vue.h)(_formItem.default, {
            title: (0, _core.getI18n)('vxe.formDesign.widgetProp.name')
          }, {
            default() {
              return (0, _vue.h)(_input.default, {
                modelValue: widget.title,
                'onUpdate:modelValue'(val) {
                  widget.title = val;
                }
              });
            }
          }), (0, _vue.h)(_formItem.default, {
            title: (0, _core.getI18n)('vxe.formDesign.widgetProp.uploadProp.multiImg'),
            field: 'multiple',
            itemRender: {
              name: 'VxeSwitch'
            }
          }), options.multiple ? (0, _vue.h)(_formItem.default, {
            title: (0, _core.getI18n)('vxe.formDesign.widgetProp.uploadProp.limitImgCount'),
            field: 'limitCount',
            itemRender: {
              name: 'VxeInput',
              props: {
                type: 'integer',
                min: 1,
                clearable: true
              }
            }
          }) : (0, _vue.createCommentVNode)(), (0, _vue.h)(_formItem.default, {
            title: (0, _core.getI18n)('vxe.formDesign.widgetProp.uploadProp.limitImgSize'),
            field: 'limitSize',
            itemRender: {
              name: 'VxeSelect',
              options: limitSizeOptions.value
            }
          }), (0, _vue.h)(_formItem.default, {
            title: (0, _core.getI18n)('vxe.formDesign.widgetProp.required')
          }, {
            default() {
              return (0, _vue.h)(_switch.default, {
                modelValue: widget.required,
                'onUpdate:modelValue'(val) {
                  widget.required = val;
                }
              });
            }
          })];
        }
      });
    };
  }
});
